tests/EZConfig.hs

```diff
@@ -30,15 +44,59 @@ spec = do
     it "Fails on the non-existent key M-10" $
       readKeySequence def "M-10" `shouldBe` Nothing
 
-regularKeys :: ([String], [KeySym])
-regularKeys = unzip . map (first (: ""))
-            $ zip ['!'    .. '~'   ] [xK_exclam       .. xK_asciitilde]
-           ++ zip ['\xa0' .. '\xff'] [xK_nobreakspace .. xK_ydiaeresis]
+-- | Parsing preserves all info that printing does.
+prop_encodePreservation :: KeyString -> Property
+prop_encodePreservation (coerce -> s) = parse s === (parse . pp =<< parse s)
+ where parse = runParser (parseKeySequence def)
+       pp    = unwords . map keyToString
+
+-- | Printing preserves all info that parsing does.
+prop_decodePreservation :: NonEmptyList (AKeyMask, AKeySym) -> Property
+prop_decodePreservation (getNonEmpty >>> coerce -> xs) =
+  Just (pp xs) === (fmap pp . parse $ pp xs)
+ where parse = runParser (parseKeySequence def)
+       pp    = unwords . map keyToString
 
 -- | QuickCheck can handle the 8! combinations just fine.
 modifiers :: [String]
-modifiers = map concat $
-  permutations ["M-", "C-", "S-", "M1-", "M2-", "M3-", "M4-", "M5-"]
+modifiers = map concat $ permutations mods
+
+mods :: [String]
+mods = ["M-", "C-", "S-", "M1-", "M2-", "M3-", "M4-", "M5-"]
 
 surround :: String -> String
 surround s = "<" <> s <> ">"
+
+-----------------------------------------------------------------------
+-- Newtypes and Arbitrary instances
+
+newtype AKeyMask = AKeyMask KeyMask
+  deriving newtype (Show)
+
+instance Arbitrary AKeyMask where
+  arbitrary :: Gen AKeyMask
+  arbitrary = fmap (coerce @Word32 . sum . nub) . listOf . elements $
+    -- 2^1 = 2 is lock; EZConfig doesn't parse this.
+    0 : [2^i | i <- [0, 2, 3, 4, 5, 6, 7] :: [Int]]
```

Comment:
Might as well just use `[shiftMask, controlMask, mod1Mask, mod2Mask, mod3Mask, mod4Mask, mod5Mask]` … :-)

Reply:
Oh, right, that's much clearer :)